src/DurableTask.Core/Common/Utils.cs

```diff
@@ -512,5 +516,111 @@ public static bool TryGetTaskScheduledId(HistoryEvent historyEvent, out int task
+
+            return ConvertFromGenericType(genericParameters: methodInfo.GetGenericArguments(), genericArguments, genericArgument);
+        }
+
+        /// <summary>
+        /// Returns the generic arguments from the invoked method if any.
+        /// </summary>
+        /// <param name="binder">The invoke member binder.</param>
+        /// <param name="methodInfo">The invoked method info.</param>
+        /// <returns>The generic arguments.</returns>
+        /// <exception cref="InvalidOperationException">Thrown when the invoked member is not C#.</exception>
+        internal static Type[] GetGenericMethodArguments(InvokeMemberBinder binder, MethodInfo methodInfo)
+        {
+            var binderType = binder.GetType()!;
+
+            if (methodInfo.IsGenericMethod && !string.Equals(binderType.Name, "CSharpInvokeMemberBinder"))
+            {
+                throw new InvalidOperationException("Generic method invoked but method binder is not C#");
+            }
+
+            Type[] genericTypeArguments = binderType.GetProperty("TypeArguments")?.GetValue(binder) as Type[] ?? Array.Empty<Type>();
+
+            if (genericTypeArguments.Length != methodInfo.GetGenericArguments().Length)
+            {
+                throw new InvalidOperationException("Generic method mismatch");
+            }
+
+            return genericTypeArguments;
+        }
+
+        /// <summary>
+        /// Converts the specified <paramref name="typeToConvert"/> to a non-generic equivalent.
+        /// </summary>
+        /// <param name="genericParameters">The generic type parameters.</param>
+        /// <param name="genericArguments">The generic type arguments.</param>
+        /// <param name="typeToConvert">The type to convert.</param>
+        /// <returns>The non-generic representation of the type.</returns>
+        internal static Type ConvertFromGenericType(Type[] genericParameters, Type[] genericArguments, Type typeToConvert)
+        {
+            if (typeToConvert.IsArray)
+            {
+                var elementType = typeToConvert.GetElementType();
+                if (elementType.IsGenericParameter)
+                {
+                    var index = Array.IndexOf(genericParameters, elementType);
+
+                    // Return the value of the generic argument.
+                    return ConvertFromGenericType(
+                        genericParameters,
+                        genericArguments,
+                        genericArguments[index].MakeArrayType());
+                }
+            }
+
+            if (typeToConvert.IsGenericType)
+            {
+                var genericArgs = typeToConvert.GetGenericArguments();
+                List<Type> genericTypeValues = new();
+
+                foreach (var genericArg in genericArgs)
+                {
+                    var index = Array.IndexOf(genericParameters, genericArg);
```

Comment:
Prefer explicit types over `var` if the right-hand expression doesn't already include the explicit type name.
```suggestion
                    int index = Array.IndexOf(genericParameters, genericArg);
```